xarray/core/variable.py

```diff
@@ -540,6 +540,20 @@ def _broadcast_indexes(self, key):
             k.item() if isinstance(k, np.ndarray) and k.ndim == 0 else k for k in key
         )
 
+        key_dict = dict(zip(self.dims, key))
+        for dim, k in key_dict.items():
+            if isinstance(k, Iterable):
+                # let da.sel(x=da.x) pass but skip if Variable has different dimensions
+                # e.g. da.sel(x=Variable(("points",), [0, 1, 2]))
+                if isinstance(k, Variable) and k.dims != (dim,):
+                    continue
+                if duck_array_ops.array_equiv(k, np.arange(self.sizes[dim])):
+                    # short-circuit when keys are effectively slice(None)
+                    # This preserves dask name and passes lazy array equivalence checks
+                    # (see duck_array_ops.lazy_array_equiv)
+                    key_dict[dim] = slice(None)
+        key = tuple(key_dict.values())
```

Comment:
This is the big change. It rewrites things like `da.sel(x=da.x)` to `da.sel(x=slice(None))` which seems to do what we want i.e. not insert a `getitem` call in the dask graph.

A nice benefit is that `sortby` on a sorted object just returns the same object.

Is this the right place to make this change? (I wasn't comfortable going any deeper into the indexing code).

Reply:
My concern is that this changes the semantics of xarray indexing on NumPy arrays, because we now return a view of NumPy arrays rather than a copy.

Reply:
hmmm.... How about we special case this section and have it run only on dask arrays. The view-vs-copy distinction does not apply to dask arrays, does it?

Reply:
I agree, this would be safe on dask arrays. But it probably makes sense to do that at a lower level, e.g., inside `DaskIndexingAdapter`. Or maybe even inside dask itself?

Reply:
Now in `DaskIndexingAdapter`